### PR TITLE
fix(controller): update replicaset annotation if needed

### LINKED ISSUE #102551: Deployment rolling update hang during scaling
#### What happened:

Deployment rolling update hangs if it is scaling at the same time.

#### What you expected to happen:

Changes in deployment's replicas do not block rolling update.

#### How to reproduce it (as minimally and precisely as possible):

Supposing we have a deployment A with replicas 5, and the maxUnavailable is 1 and maxSurge is 0, then we trigger a rolling update (e.g. update image).
1. The deployment controller create a new replicaset (RS1) and scale down the old replicaset (RS2). The replicas of RS1 is 0, and RS2 will be scaled down to 4. Besides, there will be two annotations (i.e. DesiredReplicasAnnotation and MaxReplicasAnnotation) on RS2 with value 5.
2. Before deployment controller scales up the new replicaset RS1, we manually scale down the replicas of deployment A to 4.
3. The deployment controller will only try to sync the scaling event before rolling update.
4. The scale method in `DeploymentController` will not work because the replicas of active replicaset RS2 equals to deployment A, even though the annotation is set to 5.

#### Environment:
- Kubernetes version (use `kubectl version`):
1.18.10
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by peng09**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Fixes #102551, deployment rolling update hangs during scaling

#### Which issue(s) this PR fixes:

Fixes #102551

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by peng09** [comment]
I signed CLA

**Comment by peng09** [comment]
I signed it

**Comment by peng09** [comment]
I sign it

File: pkg/controller/deployment/sync.go [hunk 1]
```diff
@@ -300,6 +300,12 @@ func (dc *DeploymentController) scale(deployment *apps.Deployment, newRS *apps.R
 	// deployment. If there is no active replica set, then we should scale up the newest replica set.
 	if activeOrLatest := deploymentutil.FindActiveOrLatest(newRS, oldRSs); activeOrLatest != nil {
 		if *(activeOrLatest.Spec.Replicas) == *(deployment.Spec.Replicas) {
+			annotationsNeedUpdate := deploymentutil.ReplicasAnnotationsNeedUpdate(activeOrLatest, *(deployment.Spec.Replicas), *(deployment.Spec.Replicas)+deploymentutil.MaxSurge(*deployment))
```
Comment (wzshiming) [inline pkg/controller/deployment/sync.go:303]: Can we add a test to cover this?

**Comment by wzshiming** [comment]
plase sign tha CLA

File: pkg/controller/deployment/sync.go (hunk 1 above)
Comment (peng09) [reply pkg/controller/deployment/sync.go:303]: Add a new test case for this, @wzshiming please take a look.

File: pkg/controller/deployment/sync_test.go [hunk 2]
```diff
@@ -315,6 +337,16 @@ func TestScale(t *testing.T) {
 			// Get all the UPDATE actions and update nameToSize with all the updated sizes.
 			for _, action := range fake.Actions() {
 				rs := action.(testclient.UpdateAction).GetObject().(*apps.ReplicaSet)
+				if annotations, ok := test.expectedOldReplicasetAnnotations[rs.Name]; ok {
```
Comment (wzshiming) [inline pkg/controller/deployment/sync_test.go:340]: It seems that with or without this patch, the added test can pass, 
Maybe take it outside of `for _, action := range fake.Actions()`

**Pull request closed without merging**